Posting some RCA:

Cause (test-code): Deterministic failure of com.facebook.presto.flightshim.TestFlightShimTpchPlugin.close in presto-flight-shim: teardown races an in-flight async GetStream task; when the Flight netty executor terminates mid-stream, ArrowBatchSource buffers aren't released and allocator.close() throws "Memory was leaked by query. Memory leaked: (4096)".
Fix: In presto-flight-shim, drain in-flight GetStream tasks before tearing down: in AbstractTestFlightShimPlugins.close() call producer.shutdown()+awaitTermination (or producer.close()) BEFORE server.close(); and/or make FlightShimProducer.runGetStreamAsync release ArrowBatchSource/VectorSchemaRoot buffers even when the Flight event executor is already terminated so the flight-shim child allocator has no leaked 4096-byte buffer. Then re-run suite 'presto-flight-shim'.
Confidence: :large_yellow_circle: medium
Owner: Bryan Cutler

Evidence
• pipeline-run-ut-suite #49676 console L13547-13554: [ERROR] TestFlightShimTpchPlugin.close ... java.lang.IllegalStateException: Memory was leaked by query. Memory leaked: (4096) Allocator(flight-shim)... at FlightShimProducer.close(FlightShimProducer.java:215) at AbstractTestFlightShimPlugins.close(:162)
• #49676 console L12829-12870 (async stream): ERROR flight-shim-0 FlightShimProducer Error getting connector flight stream; java.util.concurrent.RejectedExecutionException: event executor terminated; Suppressed: IllegalStateException: Attempting operation on allocator when allocator is closed at ArrowBatchSource.close(:106) at FlightShimProduce
• lakehouse/presto AbstractTestFlightShimPlugins.close(): @AfterClass close(): super.close(); server.shutdown(); producer.shutdown(); server.close(); producer.close(); allocator.close(); — server closed before shimExecutor drains in-flight streams (producer.shutdown() does not await)

--- a/presto-flight-shim/src/test/java/com/facebook/presto/flightshim/AbstractTestFlightShimPlugins.java
+++ b/presto-flight-shim/src/test/java/com/facebook/presto/flightshim/AbstractTestFlightShimPlugins.java
@@ -150,17 +150,16 @@ public abstract class AbstractTestFlightShimPlugins
             throws Exception
     {
         super.close();
-        if (server != null && producer != null) {
-            server.shutdown();
-            producer.shutdown();
-        }
         if (server != null) {
-            server.close();
-            server = null;
+            server.shutdown();
         }
         if (producer != null) {
             producer.close();
             producer = null;
+        }
+        if (server != null) {
+            server.close();
+            server = null;
         }
         if (allocator != null) {
             allocator.close();

